docs/book/profiler.md

```diff
@@ -0,0 +1,257 @@
+# Profiler
+
+This page describes how to use the code profiler for Rego. The Profiler
+captures the following information about Rego expressions:
+
+  * Expression Time
+  * Number of Evaluations
+  * Number of Re-Evaluations (Redo)
+  * Expression Location in source code
+
+OPA's `eval` command provides the following profiler options
+
+| Command line Option        | Detail                                                        | Default |
+|:--------------------------:|:-------------------------------------------------------------:|:-----:|
+| `profile`                  | Returns profiling results sorted by the default or given criteria              | NA |
+| `profile-sort`             | Criteria to sort the expression profiling results. This options implies `--profile`. |  total_time_ns => num_eval => num_redo => file => line |
+| `profile-limit`            | Desired number of profiling results sorted on the given criteria. This options implies `--profile`. |  10 |
+
+## Sort criteria for the profile results
+  * `total_time_ns` - Results are displayed is decreasing order of _expression evaluation time_
+  * `num_eval`  - Results are displayed is decreasing order of _number of times an expression is evaluated_
+  * `num_redo`  - Results are displayed is decreasing order of _number of times an expression is re-evaluated(redo)_
+  * `file`  - Results are sorted in reverse alphabetical order based on the _rego source filename_
+  * `line`  - Results are displayed is decreasing order of _expression line number_ in the source file
+
+When the sort criteria is not provided `total_time_ns` has the **highest** priority
+while `line` has the **lowest**.
+
+#### Example Policy
+The different profiling examples shown later on this page use the below
+sample policy.
+
+{%ace edit=true, lang='python'%}
+package rbac
+
```

Comment:
Here's a slightly refactored version of this policy that's been passed through opa fmt:

```
package rbac

# Example input request.
input = {
	"subject": "bob",
	"resource": "foo123",
	"action": "write",
}

# Example RBAC configuration.
bindings = [
	{
		"user": "alice",
		"roles": ["dev", "test"],
	},
	{
		"user": "bob",
		"roles": ["test"],
	},
]

roles = [
	{
		"name": "dev",
		"permissions": [
			{"resource": "foo123", "action": "write"},
			{"resource": "foo123", "action": "read"},
		],
	},
	{
		"name": "test",
		"permissions": [{"resource": "foo123", "action": "read"}],
	},
]

# Example RBAC policy implementation.

default allow = false

allow {
	user_has_role[role_name]
	role_has_permission[role_name]
}

user_has_role[role_name] {
	binding := bindings[_]
	binding.user = input.subject
	role_name := binding.roles[_]
}

role_has_permission[role_name] {
	role := roles[_]
	role_name := role.name
	perm := role.permissions[_]
	perm.resource = input.resource
	perm.action = input.action
}
```

Reply:
Updated the policy.